core/commands/dag/dag.go

```diff
@@ -241,3 +277,487 @@ var DagResolveCmd = &cmds.Command{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+By default, after all car files have been processed, an attempt is made to
+pin each individual root specified in the car headers, before GC runs again.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI. Defaults to true when STDERR is a TTY"),
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		silent, _ := req.Options[silentOptionName].(bool)
+		encType, _ := req.Options[cmds.EncLong].(string)
+
+		if silent || encType != "text" {
+			// force-disable progress unless on non-silent CLI
+			req.Options[progressOptionName] = false
+		} else if _, specified := req.Options[progressOptionName]; !specified {
+			// enable progress implicitly if a TTY
+			errStat, _ := os.Stderr.Stat()
+			if 0 != (errStat.Mode() & os.ModeCharDevice) {
+				req.Options[progressOptionName] = true
+			}
+		}
+
+		return nil
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		unlocker := node.Blockstore.PinLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+		showProgress, _ := req.Options[progressOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		var counts ObjectCounts
+		go importWorker(req, &res, &api, &counts, retCh)
+
+		var roots *expectedRootsSeen
+
+		if !showProgress {
+			done := <-retCh
+			if done.err != nil {
+				return done.err
+			}
+			roots = done.roots
+		} else {
+
+			progressTicker := time.NewTicker(500 * time.Millisecond)
+			defer progressTicker.Stop()
+
+		ImportLoop:
+			for {
+				select {
+				case ret := <-retCh:
+					if ret.err != nil {
+						return ret.err
+					}
+					roots = ret.roots
+
+					progressTicker.Stop()
+					if err := res.Emit(&CarImportOutput{ObjectCounts: &counts}); err != nil {
```

Comment:
This isn't safe. `Emit` can do anything it wants with the object, including send it to another goroutine.

Reply:
The reason `&counts` is a pointer ( here and your next comment ) is that this is the only way to activate `omitempty` higher up: hhttps://github.com/ipfs/go-ipfs/pull/7011/files#diff-15172926c4147422f472139313467179R73

Want to make sure you agree with the rationale before I add a bunch of mutexes.

Reply:
That's fine although I'm not sure why omitempty is important in this case. My point is that we should be copying this object safely instead of sharing it.